src/libConsensus/ConsensusLeader.cpp

```diff
@@ -63,9 +63,207 @@ bool ConsensusLeader::CheckState(Action action) {
+    subset.m_responseData.clear();
+
+    subset.m_state = m_state;
+    // add myself to subset commit map always
+    subset.m_commitPointMap.at(m_myID) = m_commitPointMap.at(m_myID);
+    subset.m_commitPoints.emplace_back(m_commitPointMap.at(m_myID));
+    subset.m_commitMap.at(m_myID) = true;
+
+    for (unsigned int j = 0; j < m_numForConsensus - 1; j++) {
+      unsigned int index = peersWhoCommitted.at(j);
+      subset.m_commitPointMap.at(index) = m_commitPointMap.at(index);
+      subset.m_commitPoints.emplace_back(m_commitPointMap.at(index));
+      subset.m_commitMap.at(index) = true;
+    }
+
+    if (DEBUG_LEVEL >= 5) {
+      LOG_GENERAL(INFO, "SubsetID: " << i);
+      for (unsigned int i = 0; i < subset.m_commitMap.size(); i++) {
+        LOG_GENERAL(INFO,
+                    "Commit map " << i << " = " << subset.m_commitMap.at(i));
+      }
+    }
+
+    random_shuffle(peersWhoCommitted.begin(), peersWhoCommitted.end());
+  }
+  // Clear out the original commit map stuff, we don't need it anymore at this
+  // point
+  m_commitPointMap.clear();
+  m_commitPoints.clear();
+  m_commitMap.clear();
+  LOG_GENERAL(INFO, "Generated " << numSubsets << " subsets of "
+                                 << m_numForConsensus
+                                 << " backups each for this consensus");
+}
+
+void ConsensusLeader::StartConsensusSubsets() {
+  LOG_MARKER();
+  lock_guard<mutex> g(m_mutex);
+
+  ConsensusMessageType type;
+  // Update overall internal state
+  if (m_state == ANNOUNCE_DONE) {
+    m_state = CHALLENGE_DONE;
+    type = ConsensusMessageType::CHALLENGE;
+  } else if (m_state == COLLECTIVESIG_DONE) {
+    m_state = FINALCHALLENGE_DONE;
+    type = ConsensusMessageType::FINALCHALLENGE;
+  }
+
+  m_numSubsetsRunning = m_consensusSubsets.size();
+  for (unsigned int index = 0; index < m_consensusSubsets.size(); index++) {
+    // If overall state has somehow transitioned from CHALLENGE_DONE or
+    // FINALCHALLENGE_DONE then it means consensus has ended and there's no
+    // point in starting another subset
+    if (m_state != CHALLENGE_DONE && m_state != FINALCHALLENGE_DONE) {
+      break;
+    }
+    ConsensusSubset& subset = m_consensusSubsets.at(index);
+    vector<unsigned char> challenge = {m_classByte, m_insByte,
+                                       static_cast<unsigned char>(type)};
+    bool result = GenerateChallengeMessage(
+        challenge, MessageOffset::BODY + sizeof(unsigned char), index);
+    if (result) {
+      // Update subset's internal state
+      SetStateSubset(index, m_state);
+
+      // Add the leader to the responses
+      Response r(*m_commitSecret, subset.m_challenge, m_myPrivKey);
+      subset.m_responseData.emplace_back(r);
+      subset.m_responseDataMap.at(m_myID) = r;
+      subset.m_responseMap.at(m_myID) = true;
+      subset.m_responseCounter = 1;
+
+      // Multicast to all nodes who send validated commits
```

Comment:
This comment now only applies to the `else` case below, so I suggest to move it there, and have a more accurate comment for the `if (BROADCAST_GOSSIP_MODE)` branch as well.